### ISSUE TITLE
[Feat]: Use three way merge

### ISSUE THREAD

**Issue by egulatee** [AndreasAugustin/actions-template-sync#583]
### Describe the feature

Consider using a three way merge to update derived projects from a template.

### Use Case

Context:  A derived project that has modified files imported from a template.  PRs opened to sync the template will suggest reverting changes back to the template's version.

This is particularly annoying if you are using the template to dictate flow (build, workflows) with derived projects making changes to suit their needs.

### Proposed Solution

Use a three way merge to identify whether the change was done in the derived project (Thus shouldn't be part of a PR) or in the template (Should be part of a PR).  

### Acknowledgements

- [ ] I may be able to implement this feature request

**Comment by AndreasAugustin** [comment]
Hi @egulatee and thanks for the feature request.
There is actually the feature of ignoring changes with a `.templatesyncignore` file. This is working like a `.gitignore` file. Is this a way to handle your issue?

**Comment by egulatee** [comment]
Hi @AndreasAugustin 

Unfortunately not.  

I've used [three way merges](https://blog.git-init.com/the-magic-of-3-way-merge/) in the past.  

Basically I have files I am inheriting from my template project and intentionally making changes in my derived projects.  I can edit the merge to remove them, but the more changes I have within a derived project to shared files, the more annoying it will become.

This would allow the sync to determine whether the changes originate from the template or the derived project.  
If they are coming from the template then they should be included in the PR.  (New changes in the template project)
If they are coming from the derived project then they should not be included in the PR.  (Someone made the change since the template and likely wants to keep the change from the derived project)

**Comment by AndreasAugustin** [comment]
Hmm 🤔 possibly a change for the git remote pull params is a possible way to go? I can imaging using `-X yours` instead of using the default could work (if I got your description right).

At the moment I see the issue with merge conflicts. So somehow it needs to be decided which version to use in all cases. 
Do you have possibly an example or an idea of commands to use for your use case?

**Comment by egulatee** [comment]
Yes, if someone makes changes to the same lines on both the template project and child project, there would be a conflict and the user would need to pick which set of changes to keep.  (That's expected).

However if there's a change in the same file in different sections, it should be able to handle that and only include the template's changes in the PR.

RE:  Example:
I can create an example (It's as simple as changing a shared file in the child project, changing another unrelated file in the template project and then doing a sync with the template project)

RE:  Commands to use:
Let me have a look.

**Comment by egulatee** [comment]
Git performs a three-way merge automatically when you use the git merge command and a fast-forward merge is not possible. Here's how it works:
Basic Merge Command:

```
git checkout <target-branch> 
git merge <source-branch> 
target-branch: The branch you want to merge into.
source-branch: The branch you want to merge from.
```

Understanding Three-Way Merge:
Git identifies the merge base, which is the most recent common ancestor of the two branches.
It compares the merge base with the current branch (target) and the branch you're merging (source).
Git combines changes from both branches into the target branch, resolving conflicts if necessary. 


<img width="727" alt="Screenshot 2024-11-01 at 7 49 52 PM" src="https://github.com/user-attachments/assets/24e1bb1f-2987-41e3-b740-8376b051f6e2">

**Comment by egulatee** [comment]
Given that there are two repositories.  I'm not sure if there's a way to do a three way merge that way.  Not sure if there's some way to identify which template project commit the derived project came from.

<img width="698" alt="Screenshot 2024-11-01 at 7 57 27 PM" src="https://github.com/user-attachments/assets/6fa895ef-fb17-484c-b633-7f7df6042cbb">

**Comment by AndreasAugustin** [comment]
The issue I see here are how to handle merge conflicts within the automation. Imagine a conflict. How to solve it and commit and keep the conflict marker and show within the PR?
Currently not sure if there is a way. Possibly an idea?

**Comment by egulatee** [comment]
I've seen merge conflicts as having both changes in the file and the end user needs to manually edit/intervene.  

https://www.atlassian.com/git/tutorials/using-branches/merge-conflicts
https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/addressing-merge-conflicts/resolving-a-merge-conflict-using-the-command-line

$ cat merge.txt
<<<<<<< HEAD
this is some content to mess with
content to append
=======
totally different content to merge later
>>>>>>> new_branch_to_merge_later

**Comment by AndreasAugustin** [comment]
Hi @egulatee just wanted to let you know that I did not have time yet too look deeper into the issue.

**Comment by kelihlodversson** [comment]
For a three way merge, the PR-branch needs to be from the last template sync (or repository creation), and not the HEAD of the target branch. That way, the user merging the pull request will be able to solve any conflicts.  

This could be seen as similar to the [vendor branch pattern](https://github.com/Stellarium/stellarium/discussions/1856).
